src/lossNotifications.js

```diff
@@ -0,0 +1,97 @@
+import {on as onEvent} from './events.js';
+import CONSTANTS from './constants.json';
+import {getStorageManager} from '../src/storageManager.js';
+import { MODULE_TYPE_PREBID } from '../src/activities/modules.js';
+
+const MODULE_NAME = 'lossNotifications';
+const storage = getStorageManager({moduleType: MODULE_TYPE_PREBID, moduleName: MODULE_NAME});
+let tidMap = {};
+let lossNotificationsEnabled = false;
+let enabledBidders = [];
+
+export function enableLossNotifications(config) {
+  const { bidderCode, beaconUrl } = config;
+  enabledBidders.push({bidderCode, beaconUrl});
+  if (lossNotificationsEnabled) return;
+  lossNotificationsEnabled = true;
+  onEvent(CONSTANTS.EVENTS.AUCTION_END, onAuctionEndHandler);
+  onEvent(CONSTANTS.EVENTS.BID_WON, onBidWonHandler);
+  onEvent(CONSTANTS.EVENTS.BID_REQUESTED, onBidRequestedHandler);
+}
+
+function onAuctionEndHandler(auctionDetails) {
+  let tid;
+  let enabledBidderData;
+
+  auctionDetails.bidderRequests.forEach(bidderRequest => {
+    const enabledBidder = enabledBidders.find(bidder => bidder.bidderCode === bidderRequest.bidderCode);
+
+    if (enabledBidder) {
+      bidderRequest.bids.forEach(bid => {
+        tid = bid.ortb2Imp.ext.tid || bid.ortb2.source.tid || bid.transactionId || null; // should we only check bid.ortb2Imp.ext.tid here?
+        enabledBidderData = {
+          bidderCode: bidderRequest.bidderCode,
+          bidderRequestId: bidderRequest.bidderRequestId
+        };
+
+        if (enabledBidder.beaconUrl) (enabledBidderData.beaconUrl = enabledBidder.beaconUrl);
+        if (tid) {
+          if (!tidMap[tid]) {
+            tidMap[tid] = [enabledBidderData];
+          } else {
+            tidMap[tid].push(enabledBidderData);
+          }
+        }
+      });
+    }
+  });
+}
+
+function onBidWonHandler(bid) {
+  const winningBidData = {
+    bidderCode: bid.bidderCode,
+    cpm: bid.cpm,
+    status: bid.status,
+    transactionId: bid.transactionId,
+  };
+
+  tidMap[winningBidData.transactionId].forEach(bidder => {
+    if (bidder.bidderCode !== winningBidData.bidderCode) {
+      const lossNotificationPayload = {
+        bidderRequestId: bidder.bidderRequestId,
+        auctionId: winningBidData.transactionId,
+        minBidToWin: winningBidData.cpm,
+        rendered: winningBidData.status === CONSTANTS.BID_STATUS.RENDERED ? 1 : 0,
+      };
+
+      if (bidder.beaconUrl) {
```

Comment:
Using the lurl, or any other url to automatically fire a response back needs to be removed. We don't want to double the network traffic from the page. We want to strongly encourage bidders to piggyback this data on a subsequent bid request.

Reply:
I agree; I'm not sure how to do it though, but if we somehow knew subsequent bid event was not ever going to happen, eg on page change, we could queue the beacons with keepalive to survive the unload event